### ISSUE TITLE
Feature Request:  Synchronized streams aka synchronized threads

### ISSUE THREAD

**Issue by Ulmo** [yt-dlp/yt-dlp#11026]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

I watch my yt downloads as they're downloading quite often, and often the sound gets downloaded much faster than the video.  I'm concerned that this is odd behavior for a live watcher.  I don't want yt-dlp downloads to be flagged due to this simple metric.  I'd rather the two (or more) streams (aka two threads) (video and audio, in my case, but could be anything) synchronize.  Usually, this would mean slightly slowing down the audio, but the slowdown factor would apply to either stream that is going slower (sometimes the video goes faster and the video needs to be slowed down).  The slowdown amount could be slowly incremented or decremented by .1 seconds when there's a slip in either direction.  The slowdown could be implemented by sleeping the current slowdown amount of time for the current slowed down stream after each block is received after its request.  For -N where it has multiple streams, the implementor can devise a way to synchronize them properly; I don't know the sequence numbering in -N, but it seems like whatever division of labor is already used could simply easily use the same synchronization method as no -N option.

A simpler approach would be to only request the same fragment # at the same time between both streams; that would be lockstep behavior, and I think would also work just as well, and probably far simpler to implement, now that I think of it.  Each stream would be paired with whatever other streams (in the case -N plural, in the case of no -N singular) that are downloading the same fragment #s, and then when it wants a new fragment, it has to check to see if the other stream also wants the same fragment.  I call this fragment wanting synchronization.  As soon as they both want the same fragment #, they can both request it.

A "purer" approach would be a synchronized time-dependent "wanting" algorithm (see prior paragraph) that digs into the streams enough to know what time stamp there is, to solve the weird case where the duration of a fragment for each stream is differing and the fragment # is not sufficient to synchronize them.  In this case, using time stamps would be superior to using fragment #s and the fragment #s can simply be ignored.  If you are up to digging into the time code for this task, by all means, but this would also require de-noising the numbers by saying "within a double fragment duration amount" for each stream for comparison purposes, since nothing in time is identical.  However, this would be purer since it would work in those oddball circumstances of strangely different fragment durations, if that ever happens.  Also, the weird jumps and imprecisions of the denoising algorithm would be fine since it would look a bit more natural.

Extra credit:  if you want to get super special and like a challenge, make a virtual "watcher" (a virtual human) that wants to slowly adapt their "watching speed" to whatever the bandwidth seems to be, and keep the virtual watching speed constant, like a real user would, only slightly adjusting when the bandwidth seems to adjust.  Pick the same increments that YT uses.  But, don't limit it to 2x speed; many plugins allow 8x or 16x speed, so I think there's no limit to this virtual watching speed.  Of course, this is while synchronizing the streams per time, but that's implied.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
For security reasons, I'm leaving the prior debug output out.  Here is the relevant section, during download:

WARNING: Starting download from the last 120 hours of the live stream since YouTube does not have data before
If you think this is wrong, please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
[download]  669.96MiB at  569.83KiB/s (00:27:05) (frag 3680/32023)
[download]  231.39MiB at  150.53KiB/s (00:27:05) (frag 5699/32023)

As you can see, one stream (which happens to be the video stream) is at fragment 3680, and the other stream (which happens to be the audio stream) is rocketing ahead at 5699 fragments.  Who watches video like this?  I think no one.  Yes, a video viewer is very happy to buffer several gigabytes of audio because why not, but every single time?  Consistently?  I think no.
```

**Comment by Ulmo** [comment]
Extra extra credit:  watch the behavior of various real life browser viewers and see how they buffer, and use those buffering methods, as the underlayment to the virtual "watching speed" layer I mentioned under "Extra credit".  Even better if you pick the buffering algorithm that matches the current browser being spoofed.

**Comment by Ulmo** [comment]
Perhaps I ought to have called this "signature adaptations", but that's opening up a sizable can of worms.  But they seem to be getting out anyway ...

**Comment by Grub4K** [comment]
This is way out of scope for yt-dlp.

It would require not only parsing the manifest and/or container data and also the codec info to know how much and how fast should be downloaded, but also rewriting the downloader code from scratch to support multiple download operations in parallel.

While wrappers that implement this are probably possible, this is too specific and complex to implement to make it into the core code
